ZenNews (#58)
zenml-io/zenml-projects: zen-news-summarization/README.md

keep up with the news without being overwhelmed. They can also help us separate 
the signal from the noise, highlighting the most important details and helping 
us identify what's worth further investigation. 

This is where **ZenNews** come into play. It offers a tool that can 
automate the summarization process and save users time and effort while 
providing them with the information they need. This can be particularly valuable 
for busy professionals or anyone who wants to keep up with the news but doesn't 
have the time to read every article in full.

# 🎯 The goal of the project

The definition of the concrete use case aside, this project aims to showcase 
some of the advantages that ZenML brings to the table. Some major points we 
would like to highlight include:

- **The ease of use**: ZenML features [a simple and clean Python 
SDK](https://docs.zenml.io/starter-guide/pipelines). As you can 
observe in this example, it is not only used to define your steps and 
pipelines but also to access/manage the resources and artifacts that you 
interact with along the way. This makes it significantly easier for our users 

> to build their applications around ZenML.

Reviewer: This is a README document written *for* our users, so should be addressing them rather than the core dev team, I think.

> - **The extensibility**: ZenML is an extendable framework. ML projects often 

Reviewer: extendable or extensible?

- **The scalability**: This is a small PoC-like example that aims to prove 
that ZenML can help you streamline your workflows and accelerate your 
development process. However, it barely scratches the surface of how you can 
improve it even further. For more information, check 
[this section](#-limitations-future-improvements-and-upcoming-changes).

# 🐍 Base installation

The **ZenNews** project is designed as a 
[PyPI package](https://pypi.org/project/zennews/)
that you can install through `pip`:

```bash
pip install zennews
```

The package comes equipped with the following set of **key** pieces:

- **The pipeline**: The [`zen_news_pipeline`](src/zennews/pipelines/zen_news_pipeline.py) 
is the main pipeline in this workflow. In total, it features three steps, 
namely `collect`, `summarize` and `report`. The first step is responsible 
for collecting articles, the second step summarizes them, and the last step 
creates a report and posts it.
- **The steps**: There is a concrete implementation for each step defined above.
  - For the `collect` step, we have the [`bbc_news_source`](src/zennews/steps/sources/bbc.py)

>   which (on default) collects the top stories from the BBC news feed and 

Reviewer: ```suggestion
  which (by default) collects the top stories from the BBC news feed and 
```

  prepares [`Article`](src/zennews/models/article.py) objects. 
  - For the `summarize` step, we have implemented [`bart_large_cnn_samsum`](src/zennews/steps/summarize/bart_large_cnn_samsum.py)

>   step. As the name suggests, it uses bart model to generate summaries. 

Reviewer: ```suggestion
  step. As the name suggests, it uses the BART model to generate summaries. 
```

  [`post_summaries`](src/zennews/steps/report/report.py) 
  step. It showcases how a generalized step can function within a ZenML 
  pipeline and uses an alerter to share the results.
- **The materializer**: As mentioned above, the steps within our pipeline are 
using the concept of `Article`s to define their input and output space. Using 
the [`ArticleMaterializer`](src/zennews/materializers/article_materializer.py), 
we can show how to handle the materialization of these artifacts 
when it comes to a data type that is not built-in.
- **The custom stack component**: The ultimate goal of `ZenNews` is to 
serve the use the direct outcomes of the pipeline. That is why we have used it
as a chance to show the extensibility of ZenML in terms of the stack components 
and implemented a [`DiscordAlerter`](src/zennews/alerter/discord_alerter.py).
- **The CLI application**: The example also includes a 
[Click](https://click.palletsprojects.com/en/8.1.x/) CLI application. 
It utilizes how easily you can use our Python SDK to build your application 
around your ZenML workflows. In order to see it in action simply execute:

  ```bash
  zennews --help 
   ```

# 🕹 Test it locally right away

> Once you installed the `zennews` package, you are ready to test it out locally 

Reviewer: ```suggestion
After installing the `zennews` package, you are ready to test it out locally 
```

right away. The following command will get the top five articles from the BBC
news feed, summarize them and present the results to you. 

> Warning: This will temporarily switch your active ZenML stack to the 
> **default** stack and when the pipeline runs, you will download the model 
> to your local machine.

```bash
zennews bbc
```

You can also parameterize this process. In order to see the possible 
parameters, please use:

```bash
zennews bbc --help
```

# 🚀 Switching to scheduled pipelines with Vertex

The potential of an application like `zennews` can be only unlocked by 
scheduling summarization pipelines instead of manually triggering them 
yourself. In order to showcase it, we will set up a fully remote GCP stack 
and use the `VertexOrchestrator` to schedule the pipeline.

## Deploy ZenML on GCP

> Before you start building the stack, you need to deploy ZenML on GCP. For more 

Reviewer: Do we need to say 'deploy ZenML server' here instead?

Author: I wanted to stay consistent with the docs. That's why I decided to avoid the `server` here.

```bash
zenml container-registry register <CONTAINER_REGISTERY_NAME> \
    --flavor=gcp \
    --uri=<REGISTRY_URI>
```

### Artifact Store

The next component on the list is a 
[GCS artifact store](https://docs.zenml.io/component-gallery/artifact-stores/gcloud-gcs). 
In order to register it, all you have to do is to provide the path to your GCS
bucket:

```bash 
zenml artifact-store register <ARTIFACT_STORE_NAME> \
    --flavor=gcp \
    --path=<PATH_TO_BUCKET> 
```

### Orchestrator

Following the artifact store, we will register a
[Vertex AI orchestrator.](https://docs.zenml.io/component-gallery/orchestrators/gcloud-vertexai)

```bash
zenml orchestrator register <ORCHESTRATOR_NAME> \
    --flavor=vertex \
    --project=<PROJECT_ID> \
    --location=<GCP_LOCATION> \
    --workload_service_account=<EMAIL_OF_YOUR_SERVICE_ACCOUNT> \
    --service_account_path=<PATH_TO_YOUR_SERVICE_ACCOUNT_KEY>
```

You need to simply provide the id of your project, the name of your GCP 
region and the service account you would like to use.

> Warning: In this version, you have to provide both the email of the service 
> account and the path to a key.json file. This interaction will be improved 

> > with the upcoming releases.

Reviewer: ```suggestion
> in upcoming releases.
```

Make sure that the service account has the proper roles for the following 
services: Cloud Functions, Cloud Scheduler, Secret Manager, Service Account,

> Storage, and Vertex AI,

Reviewer: ```suggestion
Storage, and Vertex AI.
```

This feels a bit unclear. "the proper roles" might need to be specified and defined here?

Author: I left it a little bit vague on purpose, as this might depend on the person who is running the example. For instance, when I was running it, I mostly used `admin` roles for the GCP services but others might want to avoid that.

### GCP Stack

With these four components, we are ready to establish and activate the base 
version of our GCP stack.

```bash
zenml stack register <STACK_NAME> \
    -x <SECRETS_MANAGER_NAME> \
    -c <CONTAINER_REGISTERY_NAME> \
    -a <ARTIFACT_STORE_NAME> \
    -o <ORCHESTRATOR_NAME> \
    --set
```

### Alerter 

The last component in our stack is a special case. As mentioned before
the `zennews` package already comes equipped with a custom stack component 
implementation, namely the `DiscordAlerter`. In a nutshell, it uses the 
[**discord.py**](https://discordpy.readthedocs.io/en/stable/index.html) package
to send messages via a webhook to a discord text channel. You can find the 
implementation right [here](src/zennews/alerter/discord_alerter.py).

The following sections show how we can register `DiscordAlerter` as a custom 

> flavor , create an instance of it, and update our stack.

Reviewer: ```suggestion
flavor, create an instance of it, and update our stack.
```

#### Registering the custom flavor

 All you 

> have to do to register such custom flavor is to provide the corresponding

Reviewer: ```suggestion
have to do to register such a custom flavor is to provide the corresponding
```

`zennews` pipeline. There are three possible values that you can use for the 
`schedule` option: `hourly`, `daily` (every day at 9 AM), or `weekly` (every
Monday at 9 AM).

```bash
zennews bbc --schedule daily
```

This will use your active stack (the GCP stack) and schedule your ZenNews 
pipeline.

# 📙 Limitations, future improvements and upcoming changes

Before we end this project, it is also important to talk about the limitations
we faced, the possible future improvements, and changes that are already in 
motion:

- The first limitation of ZenNews is the number of supported news sources.
As this project was initially designed as a PoC, the only supported news 
source is BBC. However, thanks to our design, it is really easy to expand this 
list by adding additional steps, which consume data and create `Article` 
objects.
- The ability to schedule pipelines through ZenML played a critical role 
within the context of this project. However, this feature has its own 
limitations. While you can create scheduled pipelines, once the pipeline and 
its schedule is created, you can not cancel or modify the behavior of this 

> scheduled pipeline. This means, if you want to cancel it, you have to do it 
> over the orchestrator yourself.

Reviewer: ```suggestion
scheduled pipeline. This means that if you want to cancel it, you have to do it 
via the orchestrator UI or interface yourself and not from within ZenML.
```

- The other limitation regarding the schedules is the format. As of now, the 
CLI application takes the user input and converts it into a cron expression.

> Any orchestrator which does not support these expressions will not applicable.

Reviewer: ```suggestion
Any orchestrator which does not support these expressions is not supported.
```
